src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs

```diff
@@ -8,98 +8,164 @@
+                    // No need to allocate anything
+                    return source;
+                }
+
+                Dictionary<string, T> clone = new(source.Count);
+
+                // TODO append .NoAllocEnumerate() here once https://github.com/NuGet/NuGet.Client/pull/5246 is available
+                foreach (var pair in source)
+                {
+                    clone[pair.Key] = cloneFunc(pair.Value);
+                }
+
+                return clone;
+            }
         }
 
         /// <summary>
-        /// Merges the content of the other runtime graph in to this runtime graph
+        /// Merges the content of two runtime graphs and returns the combined graph.
         /// </summary>
         /// <param name="other">The other graph to merge in to this graph</param>
         public static RuntimeGraph Merge(RuntimeGraph left, RuntimeGraph right)
         {
-            var runtimes = new Dictionary<string, RuntimeDescription>();
-            foreach (var runtime in left.Runtimes.Values)
+            if (left.IsEmpty)
+            {
+                return right;
+            }
+
+            if (right.IsEmpty)
             {
-                runtimes[runtime.RuntimeIdentifier] = runtime.Clone();
+                return left;
+            }
+
+            var runtimes = new Dictionary<string, RuntimeDescription>(capacity: left.Runtimes.Count + right.Runtimes.Count);
+
+            // TODO append .NoAllocEnumerate() here once https://github.com/NuGet/NuGet.Client/pull/5246 is available
+            foreach (var pair in left.Runtimes)
+            {
+                runtimes[pair.Key] = pair.Value.Clone();
             }
 
             // Merge the right-side runtimes
-            foreach (var runtime in right.Runtimes.Values)
+            // TODO append .NoAllocEnumerate() here once https://github.com/NuGet/NuGet.Client/pull/5246 is available
+            foreach (var pair in right.Runtimes)
             {
                 // Check if we already have the runtime defined
-                RuntimeDescription leftRuntime;
-                if (runtimes.TryGetValue(runtime.RuntimeIdentifier, out leftRuntime))
+                if (runtimes.TryGetValue(pair.Key, out RuntimeDescription? leftRuntime))
                 {
                     // Merge runtimes
-                    runtimes[runtime.RuntimeIdentifier] = RuntimeDescription.Merge(leftRuntime, runtime);
+                    runtimes[pair.Key] = RuntimeDescription.Merge(leftRuntime, pair.Value);
                 }
                 else
                 {
-                    runtimes[runtime.RuntimeIdentifier] = runtime;
+                    runtimes[pair.Key] = pair.Value;
                 }
             }
 
+            var supports = new Dictionary<string, CompatibilityProfile>(capacity: right.Supports.Count + left.Supports.Count);
+
             // Copy over the right ones
-            var supports = new Dictionary<string, CompatibilityProfile>();
```

Comment:
Specify the size on the dictionary to avoid resizing. In this case it's an upper bound, as `left.Supports` is filtered to only those having restore contexts. If that's uncommon, we could reduce the left's contribution to the capacity, perhaps dividing it in half.